lib/paseto.d.ts

```diff
@@ -0,0 +1,205 @@
+declare namespace Paseto {
+    interface IPasetoKey<P extends IProtocol> {
```

Comment:
There is no unified `Key` class (and similarly below, there is no unified `Protocol` class). This is intentional. The library could've been architected to use inheritance, but I considered that abstraction to introduce unnecessary complexity, and so just defined each protocol version and key version independently at the minor cost of some redundant code, but with much cleaner delineation. I'd like to keep that clarity in the typing.

As such, could specify distinct interfaces for each key? I know it may appear redundant in the short term, but I'd like to maintain that distinction throughout the code base.

Reply:
In Typescript interface can be implemented explictly or using DuckTyping so you don't need real inheritance.

If needed we could reduce the IProtocol contract when the version3 (which should be quantum crypto if theres no need to improve V2 before that).

You can also notice that this interface do exists as the `ProtocolInterface` on the php implementation
```php
/**
 * Interface ProtocolInterface
 * @package ParagonIE\Paseto
 */
interface ProtocolInterface
{
```

Reply:
The unified IPasetoKey and IProtocol are not classes ; they are interfaces (SymmetricKey implements IPasetoKey, it doesn’t inherit from them, that would be extends). They are not expected to map a "physical" javascript implementation. In practice, they’re just here to share the comments between different key classes without having to copy-paste them ; if not for that I would never had written them in the first place.

Reply:
I recognize they are interfaces, but much as I think class abstraction may bring in unnecessary complexity so I think may unnecessary type/type signature abstraction. If they're just preventing redundancy as compared to mapping the explicit code interface, then I'd rather have that redundancy and the explicit mapping.